pydra/design/tests/test_shell.py

```diff
@@ -0,0 +1,696 @@
+import os
+from pathlib import Path
+import attrs
+import pytest
+import cloudpickle as cp
+from pydra.design import shell, TaskSpec, list_fields
+from fileformats.generic import File, Directory, FsObject
+from fileformats import text, image
+from pydra.engine.specs import MultiInputObj
+
+
+def test_interface_template():
+
+    SampleInterface = shell.define("cp <in_path> <out|out_path>")
+
+    assert issubclass(SampleInterface, TaskSpec)
+    output = shell.outarg(
+        name="out_path",
+        path_template="out_path",
+        default=True,
+        type=FsObject,
+        position=2,
+    )
+    assert sorted_fields(SampleInterface) == [
+        shell.arg(name="executable", default="cp", type=str, position=0),
+        shell.arg(name="in_path", type=FsObject, position=1),
+        output,
+    ]
+    assert sorted_fields(SampleInterface.Outputs) == [output]
+    intf = SampleInterface(in_path=File.mock("in-path.txt"))
+    assert intf.executable == "cp"
+    SampleInterface(in_path=File.mock("in-path.txt"), out_path=Path("./out-path.txt"))
+    SampleInterface.Outputs(out_path=File.mock("in-path.txt"))
+
+
+def test_interface_template_w_types_and_path_template_ext():
+
+    SampleInterface = shell.define(
+        "trim-png <in_image:image/png> <out|out_image:image/png>"
```

Comment:
is there an in/out option for those command tools that overwrite the input?

Reply:
Mmm, they are a bit tricky to handle in general in the canonical form with the hash changes (so need to copied). However, we could add special syntax to handle it like

```python
SampleInterface = shell.define(
        "trim-png <modify|image:image/png>"
)
```

which could translate to a `shell.arg(name="image", copy_mode='copy', type=image.Png)` in the inputs and a `shell.out(name="image", callable=lambda image: image, type=image.Png)` in the outputs (I think that would work)

Reply:
sounds good.

Reply:
have implemented this now